### DISCUSSION TITLE
peleLMeX post-processing

### DISCUSSION THREAD

**Discussion by Alex-jian522** [AMReX-Combustion/PeleLMeX#301] [General]
Hello, I'm having difficulty using peleLMeX post-processing.
Can peleLMeX do post-processing? Can it derive some desired quantities after calculating the results, such as diffusion coefficient, etc.? Or can it only be defined in advance in "amr.derive_plot_vars = avg_pressure mag_vort mass_fractions"
Thanks for any answer!

**Comment by baperry2** [comment]
The best method is to include these or other derived variables in advance with `amr.derive_plot_vars =  avg_pressure mag_vort mass_fractions viscosity lambda diffcoeff`. But PeleLMeX can effectively do some post-processing by running in `evaluate` mode. 

To do this run with `peleLM.run_mode = evaluate` and `peleLM.evaluate_vars = diffcoeff lambda viscosity`. Also set `amr.restart = chk00200` to evaluate these quantities from the state in an existing checkpoint file, or `amr.initDataPlt = plt00200`  to evaluate these quantities from an existing plot file. Note that this capability is not as thoroughly tested and you may run into issues evaluating certain variables, especially with `amr.initDataPlt`, so double check to make sure your results look reasonable. Another strategy would be to restart from a plot file (`amr.initDataPlt = plt00200`), add the derived variables that you want to `amr.derive_plot_vars`, and run for a single timestep in normal mode (`peleLM.run_mode = normal`) to get a new plot file with the data you want.
 
For other post processing needs, the [PeleAnalysis](https://github.com/AMReX-Combustion/PeleAnalysis) repo contains a bunch of C++ codes that can efficiently do some post processing tasks, for example computing the gradient of a variable or averaging multiple plot files. 

Finally, as discussed in https://github.com/AMReX-Combustion/PeleLMeX/issues/288#issuecomment-1772956848 another strategy is to write python scripts that use `yt` to load data from plt files and then manipulate the data as you need. If you do this, you could then use Cantera to compute thermodynamic or transport properties. The thermodynamic properties and instantaneous reaction rates evaluated in this way should exactly match what is used in PeleLMeX through PelePhysics. I believe transport coefficients may differ slightly based on implementation between Cantera and PelePhysics, but depending on your needs they may be close enough.

**Reply by Alex-jian522** [reply]
Also, I have a curious question. Why should we save variables such as mass fraction, temperature, density, etc. as a complete plt file? When sometimes only the mass fraction of certain components is needed (often not known before post-processing), doesn't this increase the difficulty of file transfer? Or are there any parameters that can save variables as a single file instead of an entire plt file?

**Reply by drummerdoc** [reply]
In our experience, it turns out it is most often useful to have all the chemical components available.  Yes, it requires additional storage, but the after-run flexibility has typically outweighed that cost.  A simple way around this is to "derive" a smaller plotfile with a yt script or one of the tools in the PeleAnalysis repo, then toss the bigger plotfile.

**Comment by Alex-jian522** [comment]
Thank you very much for your helpful answer, I will come back and leave my experience after trying these methods.

**Comment by Alex-jian522** [comment]
From my experience, I prefer to use yt for post-processing. First, it can be easily installed in Anaconda. Secondly, it has many built-in functions available. Third, because of the slow file transfer speed, I don't need all the variables in the plt file. yt can export a single variable in the plt file. Finally, yt can also convert files to facilitate further post-processing on other platforms, such as matlab.
